### PR TITLE
[NPUW] Phi-3 2K accuracy fix

### PR DESCRIPTION
**Pull request by AsyaPronina**
### Details:
 - *Added `Phi3SlidingMask` pass that replaces formula :*
 ```
 1. Range(0, atten_mask_len) <= (Range(past_kv_len, atten_mask_len).T - sliding_window)
 2. Range(0, atten_mask_len) > (Range(past_kv_len, atten_mask_len).T - sliding_window)
 3. Result = 1 | 2
 ```

*to*

```
 1. Range(0, atten_mask_len) <= (position_ids.T - sliding_window)
 2. Range(0, atten_mask_len) > (Range(past_kv_len, atten_mask_len).T - sliding_window)
 3. Intermediate_result_1 = 1 | 2
 4. Range(0, atten_mask_len) <= (Range(past_kv_len, atten_mask_len).T - sliding_window)
 5. Range(0, atten_mask_len) >= past_kv_len
 6.  Intermediate_result_2 = 4 & 5
 7. Result = Intermediate_result_1 | Intermediate_result_2
 8. To clean from attention to padding tokens: Result |= !(attention_mask[past_kv_len, atten_mask_len]).T
```
*to create correct final sliding window mask for left-padded present tokens and right-padded past ones*

### Tickets:
 - *EISW-187252*

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 1]
```diff
@@ -396,6 +472,137 @@ class GemmaSlidingMask : public ov::pass::MatcherPass {
     }
 };
 
+class Phi3SlidingMask : public ov::pass::MatcherPass {
+public:
+    OPENVINO_MATCHER_PASS_RTTI("npuw::LLMCompiledModel::Phi3SlidingMask");
+
+    explicit Phi3SlidingMask() {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: it seems in this particular case `explicit` is not required.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 2]
```diff
@@ -396,6 +472,137 @@ class GemmaSlidingMask : public ov::pass::MatcherPass {
...
+        // Searching for Phi3 sliding mask pattern to extend it to work with right-padded
+        // past tokens and left-padded present tokens.
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]:
```suggestion
        // Search for the Phi3 sliding mask pattern to extend it to work with right-padded
        // past tokens and left-padded present tokens.
```

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 3]
```diff
@@ -396,6 +472,137 @@ class GemmaSlidingMask : public ov::pass::MatcherPass {
...
+        //
+        // -  "less_equal" forms a sliding window mask, more precisely, it has following expression:
+        //
+        //        row range [0,... mask_len] <= column range [current_pos_id - sliding_window_size,
+        //                                                    ...,
+        //                                                    mask_len    -    sliding_window_size]
+        //
+        //       forming, under example conditions, the mask below:
+        //        past tokens = 3
+        //        present tokens = 5 (starting with current_pos_id = 3)
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:420]: out of curiosity, how can we get 5 present tokens with 3 past tokens?

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 4]
```diff
@@ -396,6 +472,137 @@ class GemmaSlidingMask : public ov::pass::MatcherPass {
...
+        //        sliding window len = 4
+        //                    K0 K1 K2 K3 K4 K5 K6 K7
+        //                   [ 0  1  2  3  4  5  6  7 ]
+        //        Q3[ 3 - 4 ]  0  0  0  0  1  1  1  1
+        //        Q4[ 4 - 4 ]  1  0  0  0  0  1  1  1
+        //        Q5[ 5 - 4 ]  1  1  0  0  0  0  1  1
+        //        Q6[ 6 - 4 ]  1  1  1  0  0  0  0  1
+        //        Q7[ 7 - 4 ]  1  1  1  1  0  0  0  0
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:460]: I presume in the mask we should get `i`s (as -infs), not `1`s?

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:1]: If you introduced a new pass, why there are changes to the existing code? Have you also moved stuff around?
Since the changes are quite complex nowadays, I'd avoid the unnecessary diffs to focus only on the relevant ones.

**Review by dmatveev — COMMENTED** [review]
Please try to minimize the diff

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 3 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:420]: It can be done in speculative decoding case, for example
However, `3` and `5` case was described as general simple case for dynamic shapes

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 4 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:460]: Yes, but it is done one step after. So, `inf`-s will be at the places of `1`-s, there will be `Multiply` operation onto `inf` at some point. (If I got everything correctly)

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:1]: Sure, will do!

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 4 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:460]: Got it, thanks

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:1]: Fixed, thanks!

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 1 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: Fixed, thanks!

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 2 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: Fixed, thanks!

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 5]
```diff
@@ -1022,6 +1191,10 @@ ov::npuw::LLMCompiledModel::LLMCompiledModel(const std::shared_ptr<ov::Model>& m
         LOG_INFO("Two-model pipeline will be created.");
     }
 
+    LOG_DEBUG("Try patch Phi-3 sliding window mask, if it exists.");
+    patch_phi3_sliding_mask(kvcache_model);
+    ov::save_model(kvcache_model, "swa_patched.xml");
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: This shouldn't be here

Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: Fixed, thanks!
